File: State/c++/State.cpp
#include <iostream>

class TrafficLight;

class State
{
    protected:
        TrafficLight* trafficLight;

    public:
        virtual ~State() {}
        void SetTrafficLight(TrafficLight* tr);
        virtual void nextState() = 0;
        virtual void previousState() = 0;
};

class TrafficLight
{
    private:
        State* state;

    public:
        TrafficLight(State* st) : state(nullptr)
        {
            SetState(st);
        }
        void SetState(State* st);
        void nextState();
        void previousState();
};

class GreenState : public State
{
    public:
        void nextState() override;
        void previousState() override;
};

class YellowState : public State
{
    public:
        void nextState() override;
        void previousState() override;
};

class RedState : public State
{
    public:
        void nextState() override;
        void previousState() override;
};

int main()
{
    TrafficLight* trafficLight = new TrafficLight(new GreenState());

    trafficLight->nextState();
    trafficLight->nextState();
    trafficLight->nextState();
    trafficLight->previousState();
    trafficLight->previousState();
    trafficLight->previousState();

    delete trafficLight;
    
    return 0;
}

void State::SetTrafficLight(TrafficLight *tr)
{
    trafficLight = tr;
}

void TrafficLight::SetState(State *st)
{
    if (state != nullptr) delete state;
    state= st;
    state->SetTrafficLight(this);
}

void TrafficLight::nextState()
{
    state->nextState();
}

void TrafficLight::previousState()
{
    state->previousState();
}

void GreenState::nextState()
{
    std::cout << "from green to yellow\n";
    trafficLight->SetState(new YellowState());
}

void GreenState::previousState()
{
    std::cout << "green color\n";
}

void YellowState::nextState()
{
    std::cout << "from yellow to red\n";
    trafficLight->SetState(new RedState());
}

void YellowState::previousState()
{
    std::cout << "from yellow to green\n";
    trafficLight->SetState(new GreenState());
}

void RedState::nextState()
{
    std::cout << "red color\n";
}

void RedState::previousState()
{
    std::cout << "from red to yellow\n";
    trafficLight->SetState(new YellowState());
}

File: State/python/State.py
import abc


class State(metaclass=abc.ABCMeta):

    def __init__(self):
        self._traffic_light: 'TrafficLight' = None

    @abc.abstractmethod
    def next_state(self):
        pass

    @abc.abstractmethod
    def previous_state(self):
        pass


class TrafficLight:

    def __init__(self, st: State):
        self.__state = None
        self.set_state(st)

    def set_state(self, st: State):
        self.__state = st
        self.__state._traffic_light = self

    def next_state(self):
        self.__state.next_state()

    def previous_state(self):
        self.__state.previous_state()


class GreenState(State):

    def next_state(self):
        print('from green to yellow')
        self._traffic_light.set_state(YellowState())

    def previous_state(self):
        print('Green color')


class YellowState(State):

    def next_state(self):
        print('from yellow to red')
        self._traffic_light.set_state(RedState())

    def previous_state(self):
        print('from yellow to green')
        self._traffic_light.set_state(GreenState())


class RedState(State):

    def next_state(self):
        print('red color')

    def previous_state(self):
        print('from red to yellow')
        self._traffic_light.set_state(YellowState())


if __name__ == '__main__':

    traffic_light = TrafficLight(GreenState())

    traffic_light.next_state()
    traffic_light.next_state()
    traffic_light.next_state()
    traffic_light.previous_state()
    traffic_light.previous_state()
    traffic_light.previous_state()